hourly-tips/src/solution/java/org/apache/flink/training/solutions/hourlytips/HourlyTipsTableSolution.java

```diff
@@ -0,0 +1,129 @@
+
+        job.execute();
+    }
+
+    /**
+     * Create and execute the hourly tips pipeline.
+     *
+     * @return {JobExecutionResult}
+     * @throws Exception which occurs during job execution.
+     */
+    public JobExecutionResult execute() throws Exception {
+
+        // set up streaming execution environments
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);
+
+        // start the data generator
+        DataStream<TaxiFare> fareStream = env.addSource(source);
+
+        // convert the DataStream to a Table
+        Schema fareSchema =
+                Schema.newBuilder()
+                        .column("driverId", "BIGINT")
+                        .column("tip", "FLOAT")
+                        .column("startTime", "TIMESTAMP_LTZ(3)")
+                        .watermark("startTime", "startTime")
+                        .build();
+        tableEnv.createTemporaryView("fares", fareStream, fareSchema);
+
+        // find the driver with the highest sum of tips for each hour
+        Table hourlyMax =
+                tableEnv.sqlQuery(
+                        "SELECT window_end, driverId, sumOfTips"
+                                + "  FROM ("
+                                + "    SELECT *, ROW_NUMBER() OVER (PARTITION BY window_start, window_end"
+                                + "        ORDER BY sumOfTips DESC) AS rownum"
+                                + "    FROM ("
+                                + "      SELECT window_start, window_end, driverId, SUM(tip) AS sumOfTips"
+                                + "      FROM TABLE("
+                                + "        TUMBLE(TABLE fares, DESCRIPTOR(startTime), INTERVAL '1' HOUR))"
+                                + "      GROUP BY window_start, window_end, driverId"
+                                + "    )"
+                                + "  ) WHERE rownum <= 1");
+
+        // convert the query's results into a DataStream of the type expected by the tests
+        DataStream<Tuple3<Long, Long, Float>> resultsAsStreamOfTuples =
+                tableEnv.toDataStream(hourlyMax)
+                        .map(
+                                row ->
+                                        new Tuple3<>(
+                                                row.<LocalDateTime>getFieldAs("window_end")
+                                                        .atZone(ZoneId.systemDefault())
+                                                        .toInstant()
+                                                        .toEpochMilli(),
+                                                row.<Long>getFieldAs("driverId"),
+                                                row.<Float>getFieldAs("sumOfTips")))
+                        .returns(Types.TUPLE(Types.LONG, Types.LONG, Types.FLOAT));
```

Comment:
Let me offer an alternative that stays in the Table API longer and let's that handle type conversions. If the expected type was a POJO instead of `Tuple3`, the code would be even simpler as long as the types can be extracted via reflection (it also wouldn't need to rename columns).

Either way, you don't have to go this route, but it may make sense to showcase the Table <-> DataStream conversions a bit more. A generic `DataStream.map()` does, however, offer more flexibility...

```
        DataStream<Tuple3<Long, Long, Float>> resultsAsStreamOfTuples =
                tableEnv.toDataStream(
                        hourlyMax.renameColumns(
                                $("window_end").as("f0"),
                                $("driverId").as("f1"),
                                $("sumOfTips").as("f2")),
                        DataTypes.STRUCTURED(
                                Tuple3.class,
                                DataTypes.FIELD("f0", DataTypes.TIMESTAMP_LTZ(3).bridgedTo(Long.class)),
                                DataTypes.FIELD("f1", DataTypes.BIGINT()),
                                DataTypes.FIELD("f2", DataTypes.FLOAT())

                        ));
```

Reply:
I'll include this in the discussion for the exercise, but I'm not sure which solution I like best (for inclusion in the code).